chapters/05-raster-vector.qmd

```diff
@@ -1,7 +1,797 @@
+
+::: {.panel-tabset}
+
+## DimPoints
+```{julia}
+dp = DimPoints(california_raster1)
+```
+
+## Constructing from lookups
+
+DimensionalData.jl (which underpins Rasters.jl) provides easy ways to get "lookups", i.e, axis index values, from a raster.  
+
+Note that these lookups may encode intervals, points, or anything in between - so you should use shiftlocus or set to get the actual point values!  DimPoints does this for you.  
+
+But if you want to see how this can be done automatically, here you go.
+
+```{julia}
+[(x, y) for x in dims(california_raster1, X), y in dims(california_raster1, Y)]
+```
+
+You can see that this encodes the same values as `dp` in the other tab.
+
+TODO: firm up the description here and add links.
+:::
+
+@fig-rasterize-lines-polygons shows the input vector layer, the rasterization results, and the points `pnt`.
+
+```{julia}
+#| label: fig-rasterize-lines-polygons
+#| fig-cap: Examples of line and polygon rasterization 
+#| layout-ncol: 2
+#| fig-subcap: 
+#| - Line rasterization w/ `boundary=:touches`
+#| - Polygon rasterization w/ `boundary=:center`
+# Line rasterization
+fig, ax, plt = plot(california_raster1; colormap = cgrad(:Set3; categorical = true))
+lines!(ax, california_borders; color = :darkgrey, linewidth = 1)
+scatter!(ax, vec(dp); markersize = 3, color = :black)
+display(fig)
+# Polygon rasterization
+fig, ax, plt = plot(california_raster2; colormap = cgrad(:Set3; categorical = true))
+lines!(ax, california_borders; color = :darkgrey, linewidth = 1)
+scatter!(ax, vec(dp); markersize = 3, color = :black)
+fig 
+```
+
+## Spatial vectorization {#sec-spatial-vectorization}
+
+Spatial vectorization is the counterpart of rasterization (@sec-rasterization).
+It involves converting spatially continuous raster data into spatially discrete vector data such as points, lines or polygons.
+There are three standard methods to convert a raster to a vector layer, which we cover next:
+
+-   Raster to polygons (@sec-raster-to-polygons)---converting raster cells to rectangular polygons, representing pixel areas
+-   Raster to points (@sec-raster-to-points)---converting raster cells to points, representing pixel centroids
+-   Raster to contours (@sec-raster-to-contours)
+
+Let us demonstrate all three in the given order.
+
+### Raster to polygons {#sec-raster-to-polygons}
+
+Rasters.jl does not currently have a function to convert a raster to a feature collection with one polygon per pixel or cell.  This is a similar situation in Python with `rasterio`.
+
+GeometryOps.jl offers a `polygonize` function that returns a feature collection of polygons, where each feature has a `value` property that encodes the value of all pixels within that polygon.  Each polygon contains pixels with the same value.
+
+```{julia}
+fc = GO.polygonize(src_grain)
+```
+
+To load this as a `DataFrame`, we can take a shortcut and read it as GeoJSON (which is convertable to a DataFrame).  
```

Comment:
This is a bit confusing as I typically think of GeoJSON as a file format ... I realize not that it's just a JSON string .. not sure how to make describe this better

Reply:
I should drop that and just have this do the DataFrame convert - all of this will get ironed out by https://github.com/JuliaGeo/GeoInterface.jl/pull/156 anyway!